### PR TITLE
feat(slicing): Map StorageSetKey, partition pairs to Clusters

### PR DESCRIPTION
**Pull request by ayirr7**
This PR:

- Attempts to map (storage set, partition) pairs to separate clusters, in settings. This approach came up from discussion in another PR (https://github.com/getsentry/snuba/pull/3149) which attempted to partition via "override" StorageSetKeys
- Also integrates the changes from https://github.com/getsentry/snuba/pull/3170 (remove hardcoded generics metrics storage set keys). Once PR 3170 is merged, the same/copied changes will be removed from this PR.

### CODE REVIEW

File: snuba/clusters/cluster.py
Comment (onewland) [inline snuba/clusters/cluster.py:399]: nit: unnecessary newline

File: tests/clusters/test_cluster.py
Comment (onewland) [inline tests/clusters/test_cluster.py:225]: can we also test that `get_cluster` on partition 0 returns the other cluster values? (it can be in the same method)

File: snuba/clusters/cluster.py
Comment (onewland) [inline snuba/clusters/cluster.py:421]: We should probably only build this map once (e.g `_PARTITIONED_STORAGE_SET_CLUSTER_MAP` should be a global and we only assign it if it's equal to empty map at the start of the function call).

**Review by onewland — APPROVED** [review]
This is fine with me if it works for Lyn. If so, let's close out that other PR

**Comment by onewland** [comment]
> This PR:
> 
> * Also integrates the changes from [feat(slicing): Remove hardcoded generic metrics StorageSetKeys #3170](https://github.com/getsentry/snuba/pull/3170) (remove hardcoded generics metrics storage set keys). Once PR 3170 is merged, the same/copied changes will be removed from this PR.

Note that you can change the base branch of PRs so people can focus on just the changeset you care about

**Comment by onewland** [comment]
Also, assuming we get approval from Lyn, let's see what we should update in https://github.com/getsentry/snuba/blob/master/docs/source/architecture/partitioning.rst

File: snuba/datasets/storage.py
Comment (nikhars) [inline snuba/datasets/storage.py:42]: While theoretically speaking this works, from a logical perspective it is not useful since we are not partitioning storages. What we are partitioning is storage sets in Snuba. So a storage would never have any idea about partitioning. So I would suggest we remove it from here.

File: snuba/clusters/cluster.py
Comment (nikhars) [inline snuba/clusters/cluster.py:447]: If a storage set is partitioned, it would provide a different storage set key with a suffix like `_1` without passing the partition_id. Will that behavior still work with this change?

Comment (onewland) [reply snuba/clusters/cluster.py:447]: We're not using storage set key alone to route with this approach.

A single storage set can map to multiple clusters based on the (storage set key, partition ID) pair. If we go with this approach, you'll want your mapping system to pass partition_id here, there won't be a re-routing to another storage set key.

File: snuba/datasets/storage.py
Comment (ayirr7) [reply snuba/datasets/storage.py:42]: Sure, ok that makes sense

File: tests/clusters/test_cluster.py
Comment (ayirr7) [reply tests/clusters/test_cluster.py:225]: Done

File: snuba/clusters/cluster.py
Comment (lynnagara) [inline snuba/clusters/cluster.py:440]: How is this function supposed to be used? Please add a docstring as it's imported in quite a few places. Is None is supposed to be passed only for unpartitioned, and an int for partitioned? Since they are handled completely different you may want to consider splitting it into 2 functions instead.

Another idea I'll just throw out there would be to make partition_id required, and treat everything as partitioned. An unpartitioned storage set is nothing more than a storage set with 1 partition as far as most of Snuba is concerned.

Comment (lynnagara) [reply snuba/clusters/cluster.py:440]: We'll probably also need a `get_clusters(storage_set_key)` to retrieve them all in this module. There's a bunch of pieces of code (e.g. migrations, admin, etc) that will likely need to fetch all the clusters of a storage set and iterate through them.

File: snuba/settings/__init__.py
Comment (lynnagara) [inline snuba/settings/__init__.py:43]: Please provide an example of how to use in comments above this line.

File: tests/clusters/test_cluster.py
Comment (lynnagara) [inline tests/clusters/test_cluster.py:90]: nit: I'd call this `storage_set_partitions` or something

File: snuba/settings/__init__.py
Comment (lynnagara) [reply snuba/settings/__init__.py:43]: Also, can you please clarify (in this file ideally) how CLUSTERS and PARTITIONED_CLUSTERS interacts

Comment (lynnagara) [reply snuba/settings/__init__.py:43]: It's a nitpick i know but can you put PARTITIONED_CLUSTERS below CLUSTERS in this file

**Comment by lynnagara** [comment]
This approach is generally fine to me. The biggest open question I think is how this would fit with the query path ideas laid out here https://github.com/getsentry/snuba/pull/3172

File: tests/clusters/test_cluster.py
Comment (onewland) [reply tests/clusters/test_cluster.py:90]: +1 to this suggestion

File: snuba/clusters/cluster.py
Comment (ayirr7) [reply snuba/clusters/cluster.py:440]: I like the idea of making partition id required. I'll add it and see what the others also think

Comment (onewland) [reply snuba/clusters/cluster.py:440]: > Another idea I'll just throw out there would be to make partition_id required, and treat everything as partitioned. An unpartitioned storage set is nothing more than a storage set with 1 partition as far as most of Snuba is concerned. Signature of this function would then have `partition_id: int = 0`

I've thought about this and don't like the idea. Based on the way I think we should handle topologies, you'd need to map every physical partition to a storage (https://www.notion.so/sentry/Generic-Metrics-Slicing-9c206e02f92a4abca5948d2e21c6c00d#a1c689d0674c4671ba646ad98d1cf44d).

If you don't like that idea, then we can discuss in that doc or tomorrow's design meeting, but I'd prefer to leave the unpartitioned clusters alone for the time being

Comment (lynnagara) [reply snuba/clusters/cluster.py:440]: Oh no, just to clarify I am not suggesting to change PARTITIONED_CLUSTERS or CLUSTERS anything about that interface. Only to hide the difference between PARTITIONED_CLUSTERS and CLUSTERS inside this function and make them the same to the outside world. Basically everything from CLUSTERS gets treated as partition 0 in here.

Comment (ayirr7) [reply snuba/clusters/cluster.py:442]: Hm, I'm not sure I quite follow. I think if we have it so that partition id is not optional (i.e. for the default case we pass in 0 always), the function logic would have to involve deciding whether to look into CLUSTERS or PARTITIONED_CLUSTERS. Because the function doesn't know if the storage set key we passed in is one that is not partitioned (resides in CLUSTERS) or partitioned, with multiple partition ids (resides in PARTITIONED_CLUSTERS). I think that would interfere with the clear distinction between CLUSTERS and PARTITIONED_CLUSTERS, which is defined in a comment now in settings.

Comment (lynnagara) [reply snuba/clusters/cluster.py:442]: CLUSTERS vs PARTITIONED_CLUSTERS only exists for one reason - to keep settings backwards compatible and CLUSTERS unchanged for those who don't need to care about extra slices. There is no real reason for all the different parts of Snuba to keep this distinction, in fact it adds complexity everywhere.

In this current implementation our whole code base has to know about the difference between partitioned and unpartitioned clusters. What I'm proposing is we have this function decide which one to look in. The benefit of that is that no one else has to. Right now every caller of this module needs to know the difference between partitioned and unpartitioned clusters. Why not move that job into this module and make them exactly the same to all the other parts of Snuba.

Comment (onewland) [reply snuba/clusters/cluster.py:442]: so @lynnagara is your thought that we do something like

```
_SEEN_PARTITIONED_STORAGE_SET_KEYS = extract_storage_set_keys(_PARTITIONED_STORAGE_SET_CLUSTER_MAP)

def get_cluster(storage_set_key, partition_id = 0):
   if storage_set_key in _SEEN_PARTITIONED_STORAGE_SET_KEYS:
     get from partitioned_clusters
   else:
     assert partition_id == 0
     get from unpartitioned_clusters
```

I'm fine with that if you feel strongly but I aesthetically/subjectively prefer Riya's current approach -- I don't see how it pollutes the codebase more than yours given both approaches rely on the optionality of the argument in both to avoid adding `partition_id=n` everywhere.

File: docs/source/architecture/partitioning.rst
Comment (onewland) [inline docs/source/architecture/partitioning.rst:37]:
```suggestion
regular cluster definitions in ``settings.CLUSTERS``. In the ``storage_set_partitions`` field, partitioned storage
```

File: snuba/clusters/cluster.py
Comment (onewland) [reply snuba/clusters/cluster.py:442]: OK, I'd like to continue as-is, but I agree that we are invalidating the documentation (currently) on lines 444-448 of this file and it should be corrected to clearly explain the new behavior.

Comment (nikhars) [reply snuba/clusters/cluster.py:442]: Is this a blocker or can we follow up on this separately. If it is not a blocker, can we open a follow up task and get the current implementation merged? I would like to use the functionality being introduced in this PR for https://github.com/getsentry/snuba/pull/3172

Comment (ayirr7) [reply snuba/clusters/cluster.py:442]: Right, I'll update the function docstring (apologies for not getting that done earlier)

Comment (lynnagara) [reply snuba/clusters/cluster.py:442]: Not a blocker, but I think it would be a lot better if this function took a logical partition instead of a physical partition and did the mapping in one place. I don't think there's a reason we need to call `map_logical_partition_to_physical_partition()` everywhere that then calls this function instead of just doing it just once here.

Comment (ayirr7) [reply snuba/clusters/cluster.py:442]: I see, is this referring to how this code will integrate with #3172? Do you think that change (logical to physical partition mapping inside get_cluster()) should be done before this PR gets merged?

Comment (lynnagara) [reply snuba/clusters/cluster.py:442]: No, go ahead and merge, we can figure out those details later.

**Review by lynnagara — APPROVED** [review]

**Review by lynnagara — APPROVED** [review]

File: snuba/clusters/cluster.py
Comment (ayirr7) [reply snuba/clusters/cluster.py:442]: I think I understand the idea about how every call to get_cluster() now would need to "think" about whether or not it is dealing with a partitioned or unpartitioned cluster (and whether or not to pass a slice id or not). Should changing this be a follow up task? Or should it be done before merging?

File: docs/source/architecture/partitioning.rst
Comment (ayirr7) [reply docs/source/architecture/partitioning.rst]: oops, thanks for pointing that out

**Pull request merged**
